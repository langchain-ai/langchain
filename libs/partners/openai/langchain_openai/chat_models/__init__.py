from langchain_openai.chat_models.azure import AzureChatOpenAI
from langchain_openai.chat_models.base import ChatOpenAI

__all__ = ["ChatOpenAI", "AzureChatOpenAI"]
