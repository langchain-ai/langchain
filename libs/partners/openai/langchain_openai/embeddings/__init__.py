from langchain_openai.embeddings.azure import AzureOpenAIEmbeddings
from langchain_openai.embeddings.base import OpenAIEmbeddings

__all__ = ["OpenAIEmbeddings", "AzureOpenAIEmbeddings"]
