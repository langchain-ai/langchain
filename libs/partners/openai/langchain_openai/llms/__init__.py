from langchain_openai.llms.azure import AzureOpenAI
from langchain_openai.llms.base import OpenAI

__all__ = ["OpenAI", "AzureOpenAI"]
