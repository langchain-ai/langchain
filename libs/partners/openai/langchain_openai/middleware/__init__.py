"""Middleware implementations for OpenAI-backed agents."""

from .openai_moderation import OpenAIModerationError, OpenAIModerationMiddleware

__all__ = [
    "OpenAIModerationError",
    "OpenAIModerationMiddleware",
]
