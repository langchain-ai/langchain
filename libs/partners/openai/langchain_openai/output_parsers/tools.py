from langchain_core.output_parsers.openai_tools import (
    JsonOutputKeyToolsParser, JsonOutputToolsParser, PydanticToolsParser)

__all__ = ["PydanticToolsParser", "JsonOutputToolsParser", "JsonOutputKeyToolsParser"]
