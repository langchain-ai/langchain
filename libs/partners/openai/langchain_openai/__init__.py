from langchain_openai.chat_models import AzureChatOpenAI, ChatOpenAI
from langchain_openai.embeddings import AzureOpenAIEmbeddings, OpenAIEmbeddings
from langchain_openai.llms import AzureOpenAI, OpenAI

__all__ = [
    "OpenAI",
    "ChatOpenAI",
    "OpenAIEmbeddings",
    "AzureOpenAI",
    "AzureChatOpenAI",
    "AzureOpenAIEmbeddings",
]
