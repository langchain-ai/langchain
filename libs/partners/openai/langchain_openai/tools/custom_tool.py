import inspect
from collections.abc import Awaitable
from typing import Any, Callable

from langchain_core.tools import tool


def _make_wrapped_func(func: Callable[..., str]) -> Callable[..., list[dict[str, Any]]]:
    def wrapped(x: str) -> list[dict[str, Any]]:
        return [{"type": "custom_tool_call_output", "output": func(x)}]

    return wrapped


def _make_wrapped_coroutine(
    coroutine: Callable[..., Awaitable[str]],
) -> Callable[..., Awaitable[list[dict[str, Any]]]]:
    async def wrapped(*args: Any, **kwargs: Any) -> list[dict[str, Any]]:
        result = await coroutine(*args, **kwargs)
        return [{"type": "custom_tool_call_output", "output": result}]

    return wrapped


def custom_tool(*args: Any, **kwargs: Any) -> Any:
    """Decorator to create an OpenAI custom tool.

    Custom tools allow for tools with (potentially long) freeform string inputs.

    See below for an example using LangGraph:

    .. code-block:: python

        @custom_tool
        def execute_code(code: str) -> str:
            \"\"\"Execute python code.\"\"\"
            return "27"


        llm = ChatOpenAI(model="gpt-5", output_version="responses/v1")

        agent = create_react_agent(llm, [execute_code])

        input_message = {"role": "user", "content": "Use the tool to calculate 3^3."}
        for step in agent.stream(
            {"messages": [input_message]},
            stream_mode="values",
        ):
            step["messages"][-1].pretty_print()

    You can also specify a format for a corresponding context-free grammar using the
    ``format`` kwarg:

    .. code-block:: python

        from langchain_openai import ChatOpenAI, custom_tool
        from langgraph.prebuilt import create_react_agent

        grammar = \"\"\"
        start: expr
        expr: term (SP ADD SP term)* -> add
        | term
        term: factor (SP MUL SP factor)* -> mul
        | factor
        factor: INT
        SP: " "
        ADD: "+"
        MUL: "*"
        %import common.INT
        \"\"\"

        format = {"type": "grammar", "syntax": "lark", "definition": grammar}

        # highlight-next-line
        @custom_tool(format=format)
        def do_math(input_string: str) -> str:
            \"\"\"Do a mathematical operation.\"\"\"
            return "27"


        llm = ChatOpenAI(model="gpt-5", output_version="responses/v1")

        agent = create_react_agent(llm, [do_math])

        input_message = {"role": "user", "content": "Use the tool to calculate 3^3."}
        for step in agent.stream(
            {"messages": [input_message]},
            stream_mode="values",
        ):
            step["messages"][-1].pretty_print()
    """

    def decorator(func: Callable[..., Any]) -> Any:
        metadata = {"type": "custom_tool"}
        if "format" in kwargs:
            metadata["format"] = kwargs.pop("format")
        tool_obj = tool(infer_schema=False, **kwargs)(func)
        tool_obj.metadata = metadata
        tool_obj.description = func.__doc__
        if inspect.iscoroutinefunction(func):
            tool_obj.coroutine = _make_wrapped_coroutine(func)
        else:
            tool_obj.func = _make_wrapped_func(func)
        return tool_obj

    if args and callable(args[0]) and not kwargs:
        return decorator(args[0])

    return decorator
