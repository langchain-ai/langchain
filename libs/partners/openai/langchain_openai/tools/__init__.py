from langchain_openai.tools.custom_tool import custom_tool

__all__ = ["custom_tool"]
