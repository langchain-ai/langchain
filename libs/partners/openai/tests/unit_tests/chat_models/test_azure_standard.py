"""Standard LangChain interface tests"""

from typing import Tuple, Type

import pytest
from langchain_core.language_models import BaseChatModel
from langchain_core.tools import BaseTool
from langchain_tests.unit_tests import ChatModelUnitTests

from langchain_openai import AzureChatOpenAI


class TestOpenAIStandard(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> Type[BaseChatModel]:
        return AzureChatOpenAI

    @property
    def chat_model_params(self) -> dict:
        return {
            "deployment_name": "test",
            "openai_api_version": "2021-10-01",
            "azure_endpoint": "https://test.azure.com",
        }

    @pytest.mark.xfail(reason="AzureOpenAI does not support tool_choice='any'")
    def test_bind_tool_pydantic(
        self, model: BaseChatModel, my_adder_tool: BaseTool
    ) -> None:
        super().test_bind_tool_pydantic(model, my_adder_tool)

    @property
    def init_from_env_params(self) -> Tuple[dict, dict, dict]:
        return (
            {
                "AZURE_OPENAI_API_KEY": "api_key",
                "AZURE_OPENAI_ENDPOINT": "https://endpoint.com",
                "AZURE_OPENAI_AD_TOKEN": "token",
                "OPENAI_ORG_ID": "org_id",
                "OPENAI_API_VERSION": "yyyy-mm-dd",
                "OPENAI_API_TYPE": "type",
            },
            {},
            {
                "openai_api_key": "api_key",
                "azure_endpoint": "https://endpoint.com",
                "azure_ad_token": "token",
                "openai_organization": "org_id",
                "openai_api_version": "yyyy-mm-dd",
                "openai_api_type": "type",
            },
        )
