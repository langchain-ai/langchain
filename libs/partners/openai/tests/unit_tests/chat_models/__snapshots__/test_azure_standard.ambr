# serializer version: 1
# name: TestOpenAIStandard.test_serdes[serialized]
  dict({
    'id': list([
      'langchain',
      'chat_models',
      'azure_openai',
      'AzureChatOpenAI',
    ]),
    'kwargs': dict({
      'azure_endpoint': 'https://test.azure.com',
      'deployment_name': 'test',
      'disabled_params': dict({
        'parallel_tool_calls': None,
      }),
      'max_retries': 2,
      'max_tokens': 100,
      'openai_api_key': dict({
        'id': list([
          'AZURE_OPENAI_API_KEY',
        ]),
        'lc': 1,
        'type': 'secret',
      }),
      'openai_api_type': 'azure',
      'openai_api_version': '2021-10-01',
      'request_timeout': 60.0,
      'stop': list([
      ]),
      'temperature': 0.0,
      'validate_base_url': True,
    }),
    'lc': 1,
    'name': 'AzureChatOpenAI',
    'type': 'constructor',
  })
# ---
